docs/api-tutorials/custom-reporter.md

```diff
@@ -0,0 +1,113 @@
+        // prepended to the test title
+        console.log(`${this.indent()}pass: ${test.fullTitle()}`);
+      })
+      .on(RUNNER_EVENT_FAIL, (test, err) => {
+        console.log(
+          `${this.indent()}fail: ${test.fullTitle()} - error: ${err.message}`
+        );
+      })
+      .once(RUNNER_EVENT_END, () => {
+        // Base reporter has a `stats` property
+        console.log(
+          `end: ${this.stats.passes}/${this.stats.passes +
+            this.stats.failures} ok`
+        );
+      });
+  }
+
+  indent() {
+    return Array(this._indents)
+      .fill('  ')
+      .join('');
+  }
+
+  increaseIndent() {
+    this._indents++;
+  }
+
+  decreaseIndent() {
+    this._indents--;
+  }
+}
+
+module.exports = MyReporter;
+```
+
+To use this reporter, execute `mocha --reporter /path/to/my-reporter.js`.
+
+For further examples, the built-in reporter implementations are the [best place to look](<(https://github.com/mochajs/mocha/tree/master/lib/reporters)>). The [integration tests](https://github.com/mochajs/mocha/tree/master/test/reporters) may also be helpful.
+
+## Events
+
+A reporter should listen for events emitted from the `runner` (an instance of [Runner]).
+
+The event names are exported from the `constants` property of `Mocha.Runner`:
+
+| Constant                 | Event Name  | Event Arguments | Description                                                                                 |
+| ------------------------ | ----------- | --------------- | ------------------------------------------------------------------------------------------- |
+| `RUNNER_EVENT_END`       | `end`       | _(n/a)_         | All [Suite]s, [Test]s and [Hook]s have completed execution                                  |
+| `RUNNER_EVENT_FAIL`      | `fail`      | `Test`, `Error` | A [Test] has failed or thrown an exception                                                  |
+| `RUNNER_EVENT_HOOK`      | `hook`      | `Hook`          | A [Hook] is about to execute                                                                |
+| `RUNNER_EVENT_HOOK_END`  | `hook end`  | `Hook`          | A [Hook] has completed execution                                                            |
+| `RUNNER_EVENT_PASS`      | `pass`      | `Test`          | A [Test] has passed                                                                         |
+| `RUNNER_EVENT_PENDING`   | `pending`   | `Test`          | A [Test] was skipped                                                                        |
+| `RUNNER_EVENT_RETRY`     | `retry`     | `Test`, `Error` | A [Test] failed, but is about to be retried; never emitted unless `retry` option is nonzero |
+| `RUNNER_EVENT_BEGIN`     | `start`     | _(n/a)_         | Execution will begin                                                                        |
+| `RUNNER_EVENT_SUITE`     | `suite`     | `Suite`         | The [Hook]s and [Test]s within a [Suite] are about to be executed                           |
```

Comment:
Think the event names should be bookended when appropriate.

Reply:
I am not sure what you're referring to.

Reply:
For example, `RUNNER_EVENT_SUITE` and `RUNNER_EVENT_SUITE_END` are unmatched.
Better named `RUNNER_EVENT_SUITE_BEGIN`.

Reply:
- [ ] rename.  also `RUNNER_EVENT_TEST` should be `RUNNER_EVENT_TEST_BEGIN` and `RUNNER_EVENT_HOOK` should be `RUNNER_EVENT_HOOK_BEGIN`.  reference corresponding event